pkgs/test/nixpkgs-check-by-name/src/main.rs

```diff
@@ -0,0 +1,112 @@
+mod eval;
+mod references;
+mod structure;
+mod utils;
+
+use anyhow::Context;
+use clap::Parser;
+use colored::Colorize;
+use std::ffi::OsStr;
+use std::io;
+use std::path::{Path, PathBuf};
+use std::process::ExitCode;
+use structure::Nixpkgs;
+use utils::ErrorWriter;
+
+/// Program to check the validity of pkgs/by-name
+#[derive(Parser, Debug)]
+#[command(about)]
+struct Args {
+    /// Path to nixpkgs
+    nixpkgs: PathBuf,
+}
+
+fn main() -> ExitCode {
+    let args = Args::parse();
+    match check_nixpkgs(&args.nixpkgs, None, &mut ErrorWriter::new(io::stdout())) {
+        Ok(true) => {
+            eprintln!("{}", "Validated successfully".green());
+            ExitCode::SUCCESS
+        }
+        Ok(false) => {
+            eprintln!("{}", "Validation failed, see above errors".yellow());
+            ExitCode::from(1)
+        }
+        Err(e) => {
+            eprintln!("{} {:#}", "I/O error: ".yellow(), e);
+            ExitCode::from(2)
+        }
+    }
+}
+
+pub fn check_nixpkgs<W: io::Write>(
+    nixpkgs_path: &Path,
+    extra_nix_path: Option<&OsStr>,
+    writer: &mut ErrorWriter<W>,
+) -> anyhow::Result<bool> {
+    let nixpkgs_path = nixpkgs_path.canonicalize().context(format!(
+        "Nixpkgs path {} could not be resolved",
+        nixpkgs_path.display()
+    ))?;
+
+    if !nixpkgs_path.join(structure::BASE_SUBPATH).exists() {
+        eprintln!(
+            "Given Nixpkgs path does not contain a {} subdirectory, no check necessary.",
+            structure::BASE_SUBPATH
+        );
+        return Ok(true);
+    }
+
+    let nixpkgs = Nixpkgs::new(&nixpkgs_path, writer)?;
+
+    if writer.empty {
+        // Only if we could successfully parse the structure, we do the semantic checks
+        eval::check_values(writer, &nixpkgs, extra_nix_path)?;
+        references::check_references(writer, &nixpkgs)?;
+    }
+    Ok(writer.empty)
```

Comment:
I'm not sure what you mean here. All of the functions already take a writer as an argument already, and I do pass a `Vec` in the tests to make sure each individual problem gives the expected error. I guess the tests could be for individual checks for better unit-testing, though I don't think it's very necessary here, since it's not a library meant to be re-used. Only the resulting CLI that will be used, so having the integration tests in `./tests` seem good enough to me.